src/controlplane/client/authenticator.go

```diff
@@ -0,0 +1,133 @@
+package client
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/newrelic/infra-integrations-sdk/log"
+	"github.com/newrelic/nri-kubernetes/v2/internal/config"
+	v1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/transport"
+)
+
+// Authenticator provides an interface to generate a authorized round tripper.
+type Authenticator interface {
+	Authenticate(endpoint config.Endpoint) (http.RoundTripper, error)
+}
+
+type authenticator struct {
+	logger                  log.Logger
+	secretListerByNamespace map[string]v1.SecretNamespaceLister
+	inClusterConfig         *rest.Config
+}
+
+// NewAuthenticator returns an Authenticator that supports plain, bearer token and mTLS.
+func NewAuthenticator(logger log.Logger, secretListerByNamespace map[string]v1.SecretNamespaceLister, inClusterConfig *rest.Config) Authenticator {
+	return &authenticator{
+		logger:                  logger,
+		secretListerByNamespace: secretListerByNamespace,
+		inClusterConfig:         inClusterConfig,
+	}
+}
+
+// Authenticate retruns a round tripper according to the endpoint config.
+// For mTLS configuration it fetches the certificates from the secret.
+func (a authenticator) Authenticate(endpoint config.Endpoint) (http.RoundTripper, error) {
+	transportConfig := &transport.Config{}
+	tlsConfig := transport.TLSConfig{
+		Insecure: endpoint.InsecureSkipVerify,
+	}
+
+	switch {
+	case endpoint.Auth == nil:
+		a.logger.Debugf("No authentication configured for %q, connection will be attempted anonymously", endpoint.URL)
+
+	case strings.EqualFold(endpoint.Auth.Type, bearerAuth):
+		a.logger.Debugf("Using kubernetes token to authenticate request to %q", endpoint.URL)
+
+		transportConfig.BearerToken = a.inClusterConfig.BearerToken
+
+	case strings.EqualFold(endpoint.Auth.Type, mTLSAuth) && endpoint.Auth.MTLS != nil:
+		a.logger.Debugf("Using mTLS to authenticate request to %q", endpoint.URL)
+
+		certs, err := a.getTLSCertificatesFromSecret(endpoint.Auth.MTLS)
+		if err != nil {
+			return nil, fmt.Errorf("could not load TLS configuration: %w", err)
+		}
+
+		tlsConfig.CertData = certs.cert
+		tlsConfig.KeyData = certs.key
+		// CAData could be empty if insecureSkipVerify is true.
```

Comment:
should we log something or erroring if it is empty and insecureSkipVerify is not set?

Reply:
added the check here